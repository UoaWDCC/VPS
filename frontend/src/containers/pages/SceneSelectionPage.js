import React, { useContext, useEffect, useState, useRef } from "react";
import {
  useParams,
  Route,
  useRouteMatch,
  Switch,
  useHistory,
} from "react-router-dom";
import Papa from "papaparse";
import { Button } from "@material-ui/core";
import TopBar from "../../components/TopBar";
import ListContainer from "../../components/ListContainer";
import ScreenContainer from "../../components/ScreenContainer";
import SceneContext from "../../context/SceneContext";
import AuthoringToolPage from "./AuthoringTool/AuthoringToolPage";
import { usePost, usePut, useDelete } from "../../hooks/crudHooks";
import DeleteButton from "../../components/DeleteButton";
import ShareModal from "../../components/ShareModal";
import AuthoringToolContextProvider from "../../context/AuthoringToolContextProvider";
import AuthenticationContext from "../../context/AuthenticationContext";
import HelpButton from "../../components/HelpButton";
import AccessLevel from "../../enums/route.access.level";

/**
 * Page that shows the scenes belonging to a scenario.
 *
 * @container
 */
export function SceneSelectionPage({ data = null }) {
  const [isShareModalOpen, setShareModalOpen] = useState(false);
  const { scenarioId } = useParams();
  const { url } = useRouteMatch();
  const history = useHistory();
  const { scenes, currentScene, setCurrentScene, reFetch } =
    useContext(SceneContext);
  const { getUserIdToken, VpsUser } = useContext(AuthenticationContext);
  const fileInputRef = useRef(null);
  const handCSVClick = () => {
    fileInputRef.current.click();
  };
  const handleFileChange = (e) => {
    const selectedFile = e.target.files[0];
    Papa.parse(selectedFile, {
      header: true,
      skipEmptyLines: true,
      complete(results) {
        console.log(results.data);
      },
    });
  };

  // invalid name state stores the last item that had a null name, will display error message
  const [invalidNameId, setInvalidNameId] = useState("");

  /** called when the Add card is clicked */
  async function createNewScene() {
    const newScene = await usePost(
      `/api/scenario/${scenarioId}/scene`,
      {
        name: `Scene ${scenes.length}`,
      },
      getUserIdToken
    );

    setCurrentScene(newScene);
    history.push({
      pathname: `${url}/scene/${newScene._id}`,
    });
  }

  /** called when Edit button is clicked */
  async function editScene() {
    if (currentScene != null) {
      setCurrentScene(currentScene);
      history.push({
        pathname: `${url}/scene/${currentScene._id}`,
      });
    }
  }

  /** called when Delete button is clicked */
  async function deleteScene() {
    if (currentScene != null) {
      await useDelete(
        `/api/scenario/${scenarioId}/scene/${currentScene._id}`,
        getUserIdToken
      );
      setCurrentScene(null);
      reFetch();
    }
  }

  /** called when Duplicate button is clicked */
  async function duplicateScene() {
    await usePost(
      `/api/scenario/${scenarioId}/scene/duplicate/${currentScene._id}`,
      {},
      getUserIdToken
    );
    reFetch();
  }

  /** Calls backend end point to switch to the lecturer's dashboard */
  function openDashboard() {
    history.push("/dashboard");
  }

  /** called when Play button is clicked */
  function playScenario() {
    window.open(`/play/${scenarioId}`, "_blank");
  }

  /** called when user unfocuses from a scene name */
  async function changeSceneName({ target }) {
    // Prevents user from changing scene name to empty string or one of only spaces
    if (
      target.value === "" ||
      target.value === null ||
      target.value.trim() === ""
    ) {
      target.value = currentScene.name;
      setInvalidNameId(currentScene._id);
    } else {
      setInvalidNameId("");
    }
    await usePut(
      `/api/scenario/${scenarioId}/scene/${currentScene._id}`,
      {
        name: target.value,
      },
      getUserIdToken
    );
    reFetch();
  }

  useEffect(() => {
    setCurrentScene(null);
    if (reFetch) {
      reFetch();
    }
  }, []);

  return (
    <ScreenContainer vertical>
      <TopBar>
        <DeleteButton
          className={`btn top contained ${currentScene ? "" : "disabled"}  `}
          color="default"
          variant="contained"
          disabled={!currentScene}
          onClick={deleteScene}
        >
          Delete
        </DeleteButton>
        <Button
          className={`btn top contained white ${
            currentScene ? "" : "disabled"
          }  `}
          color="default"
          variant="contained"
          disabled={!currentScene}
          onClick={editScene}
        >
          Edit
        </Button>
        <Button
          className={`btn top contained white ${
            currentScene ? "" : "disabled"
          }  `}
          color="default"
          variant="contained"
          disabled={!currentScene}
          onClick={duplicateScene}
        >
          Duplicate
        </Button>
        {VpsUser.role === AccessLevel.STAFF ? (
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={openDashboard}
          >
            Dashboard
          </Button>
        ) : (
          ""
        )}
        <Button
          className="btn top contained white"
          color="default"
          variant="contained"
          onClick={playScenario}
        >
          Play
        </Button>
        <Button
          className="btn top contained white"
          color="default"
          variant="outlined"
          onClick={() => setShareModalOpen(true)}
        >
          Share
        </Button>
        <Button
          className="btn top contained white"
          color="default"
          variant="outlined"
          onClick={() => handCSVClick(true)}
        >
          Upload CSV
        </Button>
        <input
          type="file"
          ref={fileInputRef}
          accept=".csv"
          style={{ display: "none" }}
          onChange={handleFileChange}
        />

        <HelpButton />
      </TopBar>
      <ListContainer
        data={data || scenes}
        onItemSelected={setCurrentScene}
        onItemDoubleClick={editScene}
        addCard={createNewScene}
        wide
        onItemBlur={changeSceneName}
        sceneSelectionPage
        scenarioId={scenarioId}
        invalidNameId={invalidNameId}
      />
      <ShareModal
        isOpen={isShareModalOpen}
        handleClose={() => setShareModalOpen(false)}
      />
    </ScreenContainer>
  );
}

export function ScenePage() {
  const { path } = useRouteMatch();
  return (
    <Switch>
      <Route exact path={path} component={SceneSelectionPage} />
      <AuthoringToolContextProvider>
        <Route path={`${path}/scene/:sceneId`} component={AuthoringToolPage} />
      </AuthoringToolContextProvider>
    </Switch>
  );
}
