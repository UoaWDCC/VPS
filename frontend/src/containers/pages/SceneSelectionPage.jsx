import { Button, Divider, MenuItem } from "@material-ui/core";
import ListContainer from "components/ListContainer";
import Papa from "papaparse";
import { useContext, useEffect, useRef, useState } from "react";
import {
  Route,
  Switch,
  useHistory,
  useParams,
  useRouteMatch,
} from "react-router-dom";
import ContextMenu from "../../components/ContextMenu";
import HelpButton from "../../components/HelpButton";
import ScreenContainer from "../../components/ScreenContainer";
import ShareModal from "../../components/ShareModal";
import TopBar from "../../components/TopBar";
import generateUID from "../../components/newUID";
import AuthenticationContext from "../../context/AuthenticationContext";
import AuthoringToolContextProvider from "../../context/AuthoringToolContextProvider";
import SceneContext from "../../context/SceneContext";
import AccessLevel from "../../enums/route.access.level";
import { useDelete, usePatch, usePost, usePut } from "../../hooks/crudHooks";
import AuthoringToolPage from "./AuthoringTool/AuthoringToolPage";

/**
 * Page that shows the scenes belonging to a scenario.
 *
 * @container
 */
export function SceneSelectionPage({ data = null }) {
  const [isShareModalOpen, setShareModalOpen] = useState(false);
  const { scenarioId } = useParams();
  const { url } = useRouteMatch();
  const history = useHistory();
  const { scenes, currentScene, setCurrentScene, reFetch } =
    useContext(SceneContext);
  const { getUserIdToken, VpsUser } = useContext(AuthenticationContext);

  // File input is a hidden input element that is activated via a click handler
  // This allows us to have an UI button that acts like a file <input> element.
  const fileInputRef = useRef(null);
  const handCSVClick = () => {
    fileInputRef.current.click();
  };
  const handleFileChange = (e) => {
    const selectedFile = e.target.files[0];
    Papa.parse(selectedFile, {
      header: true,
      skipEmptyLines: true,
      complete: async (results) => {
        const usersData = results.data;
        usersData.forEach(async (userData) => {
          const { name, email } = userData;
          const uid = generateUID(); // Implement your UID generation logic
          const pictureURL = "null"; // Set the picture URL if available

          await usePost(
            `/api/user/`,
            {
              name,
              uid,
              email,
              pictureURL,
            },
            getUserIdToken
          );
          console.log(`User ${name} uploaded to MongoDB.`);
        });

        // add users to the scenario
        const userEmails = usersData.map((user) => user.email);
        await usePatch(
          `/api/user/assigned/${scenarioId}`,
          { userEmails },
          getUserIdToken
        );
        console.log("CSV data uploaded to MongoDB.");
      },
    });
  };

  // invalid name state stores the last item that had a null name, will display error message
  const [invalidNameId, setInvalidNameId] = useState(""); // invalidNameId is not used?

  const [contextMenuPosition, setContextMenuPosition] = useState(null);
  function handleContextMenu(event) {
    event.preventDefault();
    if (contextMenuPosition == null) {
      setContextMenuPosition({ x: event.clientX, y: event.clientY });
    } else {
      setContextMenuPosition(null);
    }
  }

  /** called when the Add card is clicked */
  async function createNewScene() {
    const newScene = await usePost(
      `/api/scenario/${scenarioId}/scene`,
      {
        name: `Scene ${scenes.length}`,
      },
      getUserIdToken
    );

    setCurrentScene(newScene);
    history.push({
      pathname: `${url}/scene/${newScene._id}`,
    });
  }

  /** called when Edit button is clicked */
  async function editScene() {
    if (currentScene != null) {
      setCurrentScene(currentScene);
      history.push({
        pathname: `${url}/scene/${currentScene._id}`,
      });
    }
  }

  /** called when Delete button is clicked */
  async function deleteScene() {
    if (currentScene != null) {
      await useDelete(
        `/api/scenario/${scenarioId}/scene/${currentScene._id}`,
        getUserIdToken
      );
      setCurrentScene(null);
      reFetch();
    }
  }

  /** called when Duplicate button is clicked */
  async function duplicateScene() {
    await usePost(
      `/api/scenario/${scenarioId}/scene/duplicate/${currentScene._id}`,
      {},
      getUserIdToken
    );
    reFetch();
  }

  /** Calls backend end point to switch to the lecturer's dashboard */
  function openDashboard() {
    history.push("/dashboard");
  }

  /** called when Play button is clicked */
  function playScenario() {
    window.open(`/play/${scenarioId}`, "_blank");
  }

  /** called when Groups button is clicked */
  function manageGroups() {
    console.log(url, scenarioId);
    history.push({
      pathname: `${url}/manage-groups`,
    });
  }

  /** called when user unfocuses from a scene name */
  async function changeSceneName({ target }) {
    // Prevents user from changing scene name to empty string or one of only spaces
    if (
      target.value === "" ||
      target.value === null ||
      target.value.trim() === ""
    ) {
      target.value = currentScene.name;
      setInvalidNameId(currentScene._id);
    } else {
      setInvalidNameId("");
    }
    await usePut(
      `/api/scenario/${scenarioId}/scene/${currentScene._id}`,
      {
        name: target.value,
      },
      getUserIdToken
    );
    reFetch();
  }

  useEffect(() => {
    setCurrentScene(null);
    if (reFetch) {
      reFetch();
    }
  }, []);

  return (
    <ScreenContainer vertical>
      <TopBar>
        {VpsUser.role === AccessLevel.STAFF ? (
          <Button
            className="btn top contained white"
            color="default"
            variant="contained"
            onClick={openDashboard}
          >
            Dashboard
          </Button>
        ) : (
          ""
        )}
        <Button
          className="btn top contained white"
          color="default"
          variant="contained"
          onClick={manageGroups}
        >
          Groups
        </Button>
        <Button
          className="btn top contained white"
          color="default"
          variant="contained"
          onClick={playScenario}
        >
          Play
        </Button>
        <Button
          className="btn top contained white"
          color="default"
          variant="outlined"
          onClick={() => setShareModalOpen(true)}
        >
          Share
        </Button>
        {VpsUser.role === AccessLevel.STAFF ? (
          <Button
            className="btn top contained white"
            color="default"
            variant="outlined"
            onClick={() => handCSVClick(true)}
          >
            Upload CSV
          </Button>
        ) : (
          ""
        )}
        <input
          type="file"
          ref={fileInputRef}
          accept=".csv"
          style={{ display: "none" }}
          onChange={handleFileChange}
        />

        <HelpButton />
      </TopBar>

      {/* On top of the action button available in the top menu bar, we also override user's rightclick context menu to offer the same functionality. */}
      <div onContextMenu={handleContextMenu}>
        {/* Scene list */}
        <ListContainer
          data={data || scenes}
          onItemSelected={setCurrentScene}
          onItemDoubleClick={editScene}
          addCard={createNewScene}
          wide
          onItemBlur={changeSceneName}
          sceneSelectionPage
          scenarioId={scenarioId}
          invalidNameId={invalidNameId}
        />

        {/* Custom context menu */}
        <ContextMenu
          position={contextMenuPosition}
          setPosition={setContextMenuPosition}
        >
          <MenuItem disabled={!currentScene} onClick={editScene}>
            Edit
          </MenuItem>
          <MenuItem disabled={!currentScene} onClick={duplicateScene}>
            Duplicate
          </MenuItem>
          <MenuItem disabled={!currentScene} onClick={deleteScene}>
            Delete
          </MenuItem>
          <MenuItem onClick={createNewScene}>Create New Scene</MenuItem>
          <Divider />
          {VpsUser.role === AccessLevel.STAFF ? (
            <MenuItem onClick={openDashboard}>Dashboard</MenuItem>
          ) : (
            ""
          )}
          <MenuItem onClick={playScenario}>Play</MenuItem>
          <MenuItem onClick={() => setShareModalOpen(true)}>Share</MenuItem>
        </ContextMenu>
      </div>
      <ShareModal
        isOpen={isShareModalOpen}
        handleClose={() => setShareModalOpen(false)}
      />
    </ScreenContainer>
  );
}

export function ScenePage() {
  const { path } = useRouteMatch();
  return (
    <Switch>
      <Route exact path={path} component={SceneSelectionPage} />
      <AuthoringToolContextProvider>
        <Route path={`${path}/scene/:sceneId`} component={AuthoringToolPage} />
      </AuthoringToolContextProvider>
    </Switch>
  );
}
