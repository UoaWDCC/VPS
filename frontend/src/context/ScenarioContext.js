import React from "react";

const ScenarioContext = React.createContext();

export default ScenarioContext;
