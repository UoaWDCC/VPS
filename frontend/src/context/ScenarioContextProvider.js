import React, { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const { reFetch } = useGet(`api/scenario`, setScenarios);

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        currentScenario,
        setCurrentScenario,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
