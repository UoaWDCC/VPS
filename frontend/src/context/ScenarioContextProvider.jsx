import { useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";

/**
 * This is a Context Provider made with the React Context API
 * ScenarioContextProvider allows access to scenario info and the refetch function
 */
export default function ScenarioContextProvider({ children }) {
  const [currentScenario, setCurrentScenario] = useLocalStorage(
    "currentScenario",
    null
  );
  const [scenarios, setScenarios] = useState();
  const [assignedScenarios, setAssignedScenarios] = useState();
  const [roleList, setRoleList] = useState();

  const { reFetch } = useGet(`api/scenario`, setScenarios);
  const { reFetch: reFetch2 } = useGet(
    `api/scenario/assigned`,
    setAssignedScenarios
  );

  const { reFetch: reFetch3 } = useGet(
    currentScenario ? `api/group/${currentScenario._id}/roleList` : null,
    currentScenario ? setRoleList : () => {}
  );

  return (
    <ScenarioContext.Provider
      value={{
        scenarios,
        setScenarios,
        reFetch,
        assignedScenarios,
        setAssignedScenarios,
        reFetch2,
        reFetch3,
        currentScenario,
        setCurrentScenario,
        roleList,
      }}
    >
      {children}
    </ScenarioContext.Provider>
  );
}
