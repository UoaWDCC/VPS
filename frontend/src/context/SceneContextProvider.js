import React, { useContext, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";

export default function SceneContextProvider({ children }) {
  const { currentScenario } = useContext(ScenarioContext);
  const [scenes, setScenes] = useState([]);
  const [currentScene, setCurrentScene] = useLocalStorage("currentScene", null);
  let getScenes = null;
  if (currentScenario) {
    getScenes = useGet(`api/scenario/${currentScenario._id}/scene`, setScenes);
  }

  return (
    <SceneContext.Provider
      value={{
        scenes,
        setScenes,
        reFetch: getScenes?.reFetch,
        currentScene,
        setCurrentScene,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
