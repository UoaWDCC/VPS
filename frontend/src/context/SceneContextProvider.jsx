import { useContext, useEffect, useRef, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import AuthenticationContext from "./AuthenticationContext";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";
import { useParams } from "react-router-dom";
import { api } from "../util/api";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import LoadingPage from "../features/status/LoadingPage";
import GenericErrorPage from "../features/status/GenericErrorPage";
import toast from "react-hot-toast";
import useVisualScene from "../features/authoring/stores/visual";
import { getScene } from "../features/authoring/scene/scene";

async function getAllScenes(user, id) {
  const res = await api.get(user, `api/scenario/${id}/scene/all`)
  return res.data.filter(Boolean);
}

function updateScenes(user, id, sceneIds) {
  api.put(
    user,
    `/api/scenario/${id}/scene/reorder`,
    { sceneIds });
}

function deleteScene(user, scenarioId, sceneId) {
  api.delete(user, `/api/scenario/${scenarioId}/scene/${sceneId}`);
};

async function saveScene(user, scenarioId, scene) {
  const components = Object.values(scene.components);
  await api.put(user, `/api/scenario/${scenarioId}/scene/${scene._id}`, { ...scene, components });
}

/**
 * This is a Context Provider made with the React Context API
 * SceneContextProvider allows access to scene info and the refetch function
 */
export default function SceneContextProvider({ children }) {
  const { setCurrentScenario } = useContext(ScenarioContext);
  const { user } = useContext(AuthenticationContext);
  const { scenarioId } = useParams();

  const queryClient = useQueryClient();

  // TODO: this is disgusting
  const scenarioQuery = useQuery({
    queryKey: ["scenario", scenarioId],
    queryFn: () => api.get(user, `api/scenario/${scenarioId}`).then(r => r.data),
    enabled: !!scenarioId,
    onSuccess: (data) => setCurrentScenario(data),
  });

  const scenesQuery = useQuery({
    queryKey: ["scenes", scenarioId],
    queryFn: () => getAllScenes(user, scenarioId),
    enabled: !!scenarioId
  });

  const reorderMutation = useMutation({
    mutationFn: (ids) => updateScenes(user, scenarioId, ids),
    onMutate: async (ids) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        return ids.map(id => prev.find(s => s._id === id));
      });
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      // queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const deleteMutation = useMutation({
    mutationFn: (id) => deleteScene(user, scenarioId, id),
    onMutate: async (id) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], prev => prev ? prev.filter(s => s._id !== id) : []);
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const saveSceneMutation = useMutation({
    mutationFn: (scene) => saveScene(user, scenarioId, scene),
    onMutate: async (scene) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        const index = prev.findIndex(s => s._id === scene._id);
        return index === -1 ? prev : prev.toSpliced(index, 1, scene);
      })
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  if (scenesQuery.isLoading) {
    return <LoadingPage text="Getting scenes..." />;
  }

  if (scenesQuery.isError) {
    console.error(scenesQuery.error);
    return <GenericErrorPage />;
  }

  return (
    <SceneContext.Provider
      value={{
        scenes: scenesQuery.data,
        reorderScenes: reorderMutation.mutate,
        saveScene: saveSceneMutation.mutate,
        deleteScene: deleteMutation.mutate,
        reFetch: scenesQuery.refetch,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
