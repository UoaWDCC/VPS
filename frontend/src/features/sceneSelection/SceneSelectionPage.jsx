import { useContext } from "react";
import { useParams } from "react-router-dom";
import AuthenticationContext from "../../context/AuthenticationContext";
import ScenarioContext from "../../context/ScenarioContext";
import SceneContext from "../../context/SceneContext";
import { useGet } from "../../hooks/crudHooks";
import AuthoringToolPage from "../authoring/AuthoringToolPage";

/**
 * Page that shows the scenes belonging to a scenario.
 *
 * @container
 */
export function SceneSelectionPage() {
  const { scenarioId } = useParams();
  const { currentScenario, setCurrentScenario } = useContext(ScenarioContext);
  useContext(SceneContext);
  const { user } = useContext(AuthenticationContext);

  // Retrieve scenario on load
  useGet(
    `api/scenario/${scenarioId}`,
    setCurrentScenario,
    true,
    !(user && (!currentScenario || currentScenario?._id != scenarioId))
  );
}

export function ScenePage() {
  return <AuthoringToolPage />;
}
