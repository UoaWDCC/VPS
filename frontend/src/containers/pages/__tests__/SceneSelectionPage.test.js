import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { SceneSelectionPage } from "../SceneSelectionPage";
import SceneContextProvider from "../../../context/SceneContextProvider";
import ScenarioContextProvider from "../../../context/ScenarioContextProvider";
import ScenarioContext from "../../../context/ScenarioContext";

const dummyScenes = [
  { id: 1, name: "test1" },
  { id: 2, name: "test2" },
];

test("Scene Selection page snapshot test", () => {
  const context = {
    currentScenario: { _id: "scenarioId" },
  };

  const component = renderer.create(
    <BrowserRouter>
      <ScenarioContext.Provider value={context}>
        <SceneContextProvider>
          <SceneSelectionPage data={dummyScenes} />
        </SceneContextProvider>
      </ScenarioContext.Provider>
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
