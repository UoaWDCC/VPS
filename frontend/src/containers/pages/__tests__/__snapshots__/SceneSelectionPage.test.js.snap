// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene Selection page snapshot test 1`] = `
<body>
  <div>
    <div
      class="colContainer"
    >
      <div
        class="topBar"
      >
        <ul
          class="leftTopBarList"
        >
          <li
            class="listItem"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined btn top outlined white"
              href="/"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                Back
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </li>
        </ul>
        <ul
          class="rightTopBarList"
        >
          <li
            class="listItem"
          >
            <button
              class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-1 MuiButton-contained btn top contained   "
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Delete
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained btn top contained white   "
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Edit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained btn top contained white   "
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Duplicate
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained btn top contained white"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Play
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined btn top contained white"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Share
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined btn outlined white top"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
                    />
                  </svg>
                </span>
                Help
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
        </ul>
      </div>
      <div />
      <div
        class="scenarioListContainerWide"
      >
        <ul
          class="MuiImageList-root"
          style="margin: -15px;"
        >
          <li
            class="MuiImageListItem-root makeStyles-listContainerItem-2"
            height="200"
            style="width: 20%; height: 240px; padding: 15px;"
          >
            <div
              class="MuiImageListItem-item"
            >
              <div
                class="imageListItemWide"
              >
                <div
                  class="MuiBox-root MuiBox-root-3"
                />
                <p
                  class="text"
                >
                  Create New Scene
                </p>
              </div>
            </div>
          </li>
          <li
            class="MuiImageListItem-root makeStyles-listContainerItem-2"
            height="200"
            style="width: 20%; height: 240px; padding: 15px;"
          >
            <div
              class="MuiImageListItem-item"
            >
              <div
                class="imageListItemWide"
              >
                <div
                  class="MuiBox-root MuiBox-root-4"
                >
                  <div
                    data-testid="div-outer-wrapper"
                    style="display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
                  >
                    <div
                      data-testid="div-iframe-wrapper"
                      style="max-width: 266px; max-height: 150px; background-color: white; filter: drop-shadow(1px 1px 2px grey);"
                    >
                      <div
                        style="position: absolute; width: 100%; height: 150px; max-width: 266px; z-index: 10;"
                      >
                         
                      </div>
                      <span>
                        <span
                          class="react-loading-skeleton css-bvijti-skeletonStyles-Skeleton"
                          style="position: absolute; z-index: 15; width: 266px; height: 150px;"
                        >
                          ‌
                        </span>
                      </span>
                      <iframe
                        sandbox="allow-scripts allow-same-origin"
                        scrolling="no"
                        src="/play/undefined/1"
                        style="transform: scaleX(0.13854166666666667) scaleY(0.1388888888888889); transform-origin: 0 0; width: 1920px; height: 1080px; border: 0px;"
                        tabindex="-1"
                        title="webpage-thumbnail"
                      />
                    </div>
                  </div>
                </div>
                <input
                  class="text"
                  value="test1"
                />
              </div>
            </div>
          </li>
          <li
            class="MuiImageListItem-root makeStyles-listContainerItem-2"
            height="200"
            style="width: 20%; height: 240px; padding: 15px;"
          >
            <div
              class="MuiImageListItem-item"
            >
              <div
                class="imageListItemWide"
              >
                <div
                  class="MuiBox-root MuiBox-root-5"
                >
                  <div
                    data-testid="div-outer-wrapper"
                    style="display: flex; align-items: center; justify-content: center; width: 100%; height: 100%;"
                  >
                    <div
                      data-testid="div-iframe-wrapper"
                      style="max-width: 266px; max-height: 150px; background-color: white; filter: drop-shadow(1px 1px 2px grey);"
                    >
                      <div
                        style="position: absolute; width: 100%; height: 150px; max-width: 266px; z-index: 10;"
                      >
                         
                      </div>
                      <span>
                        <span
                          class="react-loading-skeleton css-bvijti-skeletonStyles-Skeleton"
                          style="position: absolute; z-index: 15; width: 266px; height: 150px;"
                        >
                          ‌
                        </span>
                      </span>
                      <iframe
                        sandbox="allow-scripts allow-same-origin"
                        scrolling="no"
                        src="/play/undefined/2"
                        style="transform: scaleX(0.13854166666666667) scaleY(0.1388888888888889); transform-origin: 0 0; width: 1920px; height: 1080px; border: 0px;"
                        tabindex="-1"
                        title="webpage-thumbnail"
                      />
                    </div>
                  </div>
                </div>
                <input
                  class="text"
                  value="test2"
                />
              </div>
            </div>
          </li>
        </ul>
      </div>
      <div />
    </div>
  </div>
</body>
`;
